compiler/rustc_typeck/src/check_unused.rs

```diff
@@ -16,48 +16,37 @@ pub fn check_crate(tcx: TyCtxt<'_>) {
         used_trait_imports.extend(imports.iter());
     }
 
-    for id in tcx.hir().items() {
-        if matches!(tcx.def_kind(id.def_id), DefKind::Use) {
-            if tcx.visibility(id.def_id).is_public() {
-                continue;
-            }
-            let item = tcx.hir().item(id);
-            if item.span.is_dummy() {
-                continue;
-            }
-            if let hir::ItemKind::Use(path, _) = item.kind {
-                check_import(tcx, &mut used_trait_imports, item.item_id(), path.span);
-            }
+    for &id in tcx.maybe_unused_trait_imports(()) {
+        debug_assert!(tcx.def_kind(id) == DefKind::Use);
+        if tcx.visibility(id).is_public() {
+            continue;
         }
+        if used_trait_imports.contains(&id) {
+            continue;
+        }
+        let item = tcx.hir().expect_item(id);
+        // if item.span.is_dummy() {
+        //     continue;
+        // }
```

Comment:
I guess that the point of this check is to avoid reporting for `use` items that are injected by macros or the compiler. You can leave it as-is, it's very fast.

Reply:
Could I add a test with a macro that injects an unused trait `use`? I'm not sure how to check the compiler-injected ones though.

Reply:
You'd need to write a proc macro for this.  The proc macro would have to create a `use Stuff;` item, where the `use` and `;` have a dummy span, and `Stuff` has a real one from the macro's input code.

Reply:
@Elliot-Roberts crafting the test can wait a follow-up PR.  If you don't mind uncommenting this code, the PR is good to merge.